src/cplscheme/config/CouplingSchemeConfiguration.cpp

```diff
@@ -948,15 +949,19 @@ void CouplingSchemeConfiguration::addDataToBeExchanged(
 
     const bool requiresInitialization = exchange.requiresInitialization;
     if (from == accessor) {
-      scheme.addDataToSend(exchange.data, exchange.mesh, requiresInitialization);
+      time::PtrWaveform ptrWaveform(new time::Waveform(_config.extrapolationOrder));
+      // @todo store all waveforms in _waveforms? See acceleration config.
+      scheme.addDataToSend(exchange.data, ptrWaveform, exchange.mesh, requiresInitialization);
```

Comment:
Couldn't we create the waveforms closer to the time-stepping part of the config? At some point, we will also get configuration options there. Here, it feels a bit misplaced.

Reply:
I would rather move this at a later point in time, if necessary. Currently adding the waveform at a different point requires us to add more functions to the `CouplingScheme` (`addWaveformToReceive`, `addWaveformToSend`) and I don't see a clear benefit.

Right now adding everything with a single function enforces the rule "each data has a waveform". This rule might be removed at a later point in time, but currently we are following this rule (because also extrapolation is applied globally to all data).

Reply:
I started drafting the shift of the ownership of the Waveform into `Data`. This is also a possible way to go, but still needs some refinement. See https://github.com/BenjaminRodenberg/precice/pull/4.

However, I think it automatically solves [this problem](https://github.com/BenjaminRodenberg/precice/pull/1/files#r745679418).